python/lsst/daf/butler/registry/collections/nameKey.py

```diff
@@ -58,167 +47,25 @@
-    @property
-    def key(self) -> str:
-        # Docstring inherited from CollectionRecord.
-        return self.name
-
-    @property
-    def host(self) -> Optional[str]:
-        # Docstring inherited from RunRecord.
-        return self._host
-
-    @property
-    def timespan(self) -> Timespan[astropy.time.Time]:
-        # Docstring inherited from RunRecord.
-        return self._timespan
-
-
-class NameKeyChainedCollectionRecord(ChainedCollectionRecord):
-    """A `ChainedCollectionRecord` implementation that just uses the string
-    name as the primary/foreign key for collections.
-    """
-    def __init__(self, db: Database, name: str, *, table: sqlalchemy.schema.Table):
-        super().__init__(name=name)
-        self._db = db
-        self._table = table
-
-    @property
-    def key(self) -> str:
-        # Docstring inherited from CollectionRecord.
-        return self.name
-
-    def _update(self, manager: CollectionManager, children: CollectionSearch):
-        # Docstring inherited from ChainedCollectionRecord.
-        rows = []
-        i = 0
-        for child, restriction in children.iter(manager, withRestrictions=True, flattenChains=False):
-            if restriction.names is ...:
-                rows.append({"parent": self.key, "child": child.key, "position": i,
-                             "dataset_type_name": ""})
-                i += 1
-            else:
-                for name in restriction.names:
-                    rows.append({"parent": self.key, "child": child.key, "position": i,
-                                 "dataset_type_name": name})
-                    i += 1
-        with self._db.transaction():
-            self._db.delete(self._table, ["parent"], {"parent": self.key})
-            self._db.insert(self._table, *rows)
-
-    def _load(self, manager: CollectionManager) -> CollectionSearch:
-        # Docstring inherited from ChainedCollectionRecord.
-        sql = sqlalchemy.sql.select(
-            [self._table.columns.child, self._table.columns.dataset_type_name]
-        ).select_from(
-            self._table
-        ).where(
-            self._table.columns.parent == self.key
-        ).order_by(
-            self._table.columns.position
-        )
-        # It's fine to have consecutive rows with the same collection name
-        # and different dataset type names - CollectionSearch will group those
-        # up for us.
-        children = []
-        for row in self._db.query(sql):
-            key = row[self._table.columns.child]
-            restriction = row[self._table.columns.dataset_type_name]
-            if not restriction:
-                restriction = ...  # we store ... as "" in the database
-            record = manager[key]
-            children.append((record.name, restriction))
-        return CollectionSearch.fromExpression(children)
-
-
-class AggressiveNameKeyCollectionManager(CollectionManager):
+class AggressiveNameKeyCollectionManager(DefaultCollectionManager):
```

Comment:
It's probably best to either drop "Aggressive" in the name here, or add it to both `DefaultCollectionManager` and `SynthIntCollectionManager`, since it's a property they all have in common (and we're not yet at the point where that change would break existing repos).

Reply:
Agree. I was thinking about implementing different pre-fetching/caching policies in the DefaultCollectionManager, than we can instantiate various managers by either providing different subclasses which just pass correct parameters to a DefaultCollectionManager constructor or maybe adding configuration options for managers in YAML.

Reply:
I renamed AggressiveNameKeyCollectionManager to NameKeyCollectionManager. My current thinking is that we'll probably have just two different classes that define different schema for collection tables (one for name keys, another for int keys) but runtime behavior (caching/pre-fetching) of these classes would be controlled by the options that we pass in the configuration (need to extend it to support this feature).